x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/AESKeyUtils.java

```diff
@@ -0,0 +1,76 @@
+
+import javax.crypto.Cipher;
+import javax.crypto.SecretKey;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+import javax.crypto.spec.SecretKeySpec;
+import java.nio.charset.StandardCharsets;
+import java.security.GeneralSecurityException;
+import java.security.Key;
+import java.util.Base64;
+
+public final class AESKeyUtils {
+    public static final int KEY_LENGTH_IN_BYTES = 32; // 256-bit AES key
+    public static final int WRAPPED_KEY_LENGTH_IN_BYTES = KEY_LENGTH_IN_BYTES + 8; // https://www.ietf.org/rfc/rfc3394.txt section 2.2
+    // parameter for the KDF function, it's a funny and unusual iter count larger than 60k
+    private static final int KDF_ITER = 61616;
+    // the KDF algorithm that generate the symmetric key given the password
+    private static final String KDF_ALGO = "PBKDF2WithHmacSHA512";
+    // The Id of any AES SecretKey is the AES-Wrap-ciphertext of this fixed 32 byte wide array.
+    // Key wrapping encryption is deterministic (same plaintext generates the same ciphertext)
+    // and the probability that two different keys map the same plaintext to the same ciphertext is very small
+    // (2^-256, much lower than the UUID collision of 2^-128), assuming AES is indistinguishable from a pseudorandom permutation.
+    private static final byte[] KEY_ID_PLAINTEXT = "wrapping known text forms key id".getBytes(StandardCharsets.UTF_8);
+
+    public static byte[] wrap(SecretKey wrappingKey, SecretKey keyToWrap) throws GeneralSecurityException {
+        assert "AES".equals(wrappingKey.getAlgorithm());
+        assert "AES".equals(keyToWrap.getAlgorithm());
+        Cipher c = Cipher.getInstance("AESWrap");
+        c.init(Cipher.WRAP_MODE, wrappingKey);
+        return c.wrap(keyToWrap);
+    }
+
+    public static SecretKey unwrap(SecretKey wrappingKey, byte[] keyToUnwrap) throws GeneralSecurityException {
+        assert "AES".equals(wrappingKey.getAlgorithm());
+        assert keyToUnwrap.length == WRAPPED_KEY_LENGTH_IN_BYTES;
+        Cipher c = Cipher.getInstance("AESWrap");
+        c.init(Cipher.UNWRAP_MODE, wrappingKey);
+        Key unwrappedKey = c.unwrap(keyToUnwrap, "AES", Cipher.SECRET_KEY);
+        return new SecretKeySpec(unwrappedKey.getEncoded(), "AES"); // make sure unwrapped key is "AES"
+    }
+
+    /**
+     * Computes the ID of the given AES {@code SecretKey}.
+     * The ID can be published as it does not leak any information about the key.
+     * Different {@code SecretKey}s have different IDs with a very high probability.
+     * <p>
+     * The ID is the ciphertext of a known plaintext, using the AES Wrap cipher algorithm.
+     * AES Wrap algorithm is deterministic, i.e. encryption using the same key, of the same plaintext, generates the same ciphertext.
+     * Moreover, the ciphertext reveals no information on the key, and the probability of collision of ciphertexts given different
+     * keys is statistically negligible.
+     */
```

Comment:
I'm going hold off reviewing the tests until we reach a conclusion on this topic ^^
If we decide to change something here that will mean I need to do another pass anyway, and even if we decide to keep this bit as-is, I think I need to get my head around it before can review the tests well.

Reply:
These are good questions, and I've spent **a lot** of thinking in this area. I believe I have good answers to them.

Encrypted (wrapped) DEKs are named after the Id of the KEK that was used to wrap them (in order to use a DEK, you need to know the KEK to unwrap it; we're aiming for possibly multiple living KEKs in a given repository, so the KEK Id is a way to identify which of the KEKs (you have already obtained using PBKDF2 on the node) to use). The crucial bit here is that the KEK Id is public, as anyone that can list the repo has access to KEK Ids. In order to avoid introducing other repo settings, the only way to derive an Id for a KEK is the content of the KEK itself (i.e. we could introduce a salt as a repo setting, in the cluster state, but then when the repo is installed on a different cluster, the password, as well as the salt, must be configured).

So, the problem statement is that we have a secret AES Key (a string of bytes), generated using PBKDF2, for which we need to generate a public identifier, given no extra input.
I would say there are 3 good (engineering-grade) reasons to use AESWrap:

- crypto gurus advise against reusing the same key for different crypto operations; in this case the KEK is already used for AESWrap of DEKs.
- AESWrap is deterministic.
- there are various trusted sources that state that knowing the plaintext and the ciphertext does not reveal any information about the AES key (regardless of mode). We actually rely on this fact in other places too, because **some repo files have predictable contents**. AESWrap inherits this property, so that even if the ciphertext (KEK Id) and the plaintext (hardcoded byte string) are known, no information is exposed on the KEK (wrapping key).

Reply:
Let me sit with this a bit longer.

I'm persuaded that:

1. If we use the KEK (as a key) to generate the ID then we should favour AESWrap due to wanting to restrict it to a single operation.
2. Just hashing (SHA) the KEK is a bad idea because that leaks something about the key (only really an oracle to tell whether a predicted key is the real key, but that's still something).

It still wonder whether there's a better option, but I'm not sure what it would be.

- We could HMAC the KEK, but what would we use as the key in the HMAC?
- We could SHA(AESWrap(KEK, ID_TEXT)), which reduces the plaintext risks, but I'm not sure if it's actually worth it, or whether it's just throwing in more operations "for the feels".

@jkakavas I'd appreciate any thoughts you have.

Reply:
> I'm persuaded that:
> If we use the KEK (as a key) to generate the ID then we should favour AESWrap due to wanting to restrict it to a single operation.

I'm not, I think. While it is true that there are generic warnings around this in bibliography, this stems from specific vulnerabilities in sets of algorithms from key reuse ( Same RSA private key for encryption and signing, AEC-CBC for encryption plus AES-CBC-MAC ) and is not a universal truth. Given that we control the algorithms used, I can see us coming down to a safe algorithm pair.


> Just hashing (SHA) the KEK is a bad idea because that leaks something about the key (only really an oracle to tell whether a predicted key is the real key, but that's still something).

I don't see the difference here. Given that the magic bytes that constitute the to-be-wrapped-key are stored in the source code, what stops a potential attacker from using AESWrap as an oracle for valid wrapping keys ( KEKs )? Also, assuming a malicious user has access to IDs, they have access to the repository and to DEKs, so they could also use unwrapping of DEKs as an oracle. Granted, a hash-and-check oracle is much more cost-efficient than a try-to-decrypt one. 


Alternative to the already proposed solutions, we could also use something like [HKDF](https://tools.ietf.org/html/rfc5869) with a null salt to get two keys from the KEK. One for AESWrap of DEKs and another for a HMAC to produce the id of the KEK.

Reply:
Thanks for thinking this through, folks!

I'm going to pile on my thoughts to maintain this as proposed:

There's indeed the risk that the contents of the encrypted repository on the cloud service can be used to brute force the keys. But, given the nature of the encrypted files, certain files or at least the headers for these files are known. We don't have any protection against this, besides making it expensive to generate the KEK (PBKDF2 cost factor). There will always be the option to brute force 256-bit AES keys.

I think we can all agree that a hash function will be more appropriate in this case. So take the properties of a cryptographic hash function:

- It's hard to find two inputs with the same output, called a collision
- It's hard to find an input matching a given output, called a pre-image

Considering the AES-Wrap encryption with the KEK key as the input (and the ciphertext of the known plaintext as the output) , the properties above are trivially met by AES-Wrap otherwise it would be vulnerable to fatal key recovery attacks.